### PR TITLE
[jsonrpc] get available-but-not-assigned artwork

### PR DESCRIPTION
**Pull request by rmrector**
## Description
Get available-but-unassigned artwork for video / music library items via JSON-RPC. This is a list of art currently available only in the GUI "Choose Art" dialog.

I added a 'preview' URL to the ScraperUrl which will be picked up from scraper / NFO XML if provided - though I didn't change the GUI to use it to keep this slim. I did not expose any of the other fields of a ScraperUrl (spoof, post, isgz, cache) as they are unnecessary or inappropriate - I'd try to remove them altogether if I weren't aiming for a narrow change for potential inclusion in Matrix.

The "arttype" parameter is optional and if omitted will return all potential images for every artwork type.

## Motivation and Context
Give a remote interface access to a feature currently available only to the GUI.

## How Has This Been Tested?
Manual testing.

## Examples

"Get art types" request and result

```json
{"jsonrpc":"2.0","id":10,"method":"VideoLibrary.GetAvailableArtTypes","params":{"item":{"tvshowid":836}}}

"availablearttypes": [
  "fanart",
  "poster",
  "banner"
]
```

"Get art" request and result

```json
{"jsonrpc":"2.0","id":12,"method":"VideoLibrary.GetAvailableArt","params":{"item":{"tvshowid":836},"arttype":"poster"}}

"availableart": [
  {
    "arttype": "poster",
    "url": "image://https%3a%2f%2fartworks.thetvdb.com%2fbanners%2fseri…"
  },
  {
    "arttype": "poster",
    "url": "image://https%3a%2f%2fartworks.thetvdb.com%2fbanners%2fseri…"
  }
]
```

## Types of change
- [X] **Improvement** (non-breaking change which improves existing functionality)

### CODE REVIEW

**Comment by DaveTBlake** [comment]
Exposing the scraped possible remote art to the API seems reasonable, will both make it possible for apps to mimic the art selection GUI as required and allow addons to make use of the previously scraped art results without reperforming matching requests at the remote sites. Thanks @rmrector.

I am also up for having this addition in v19 providing it can be completed in time, but I have a suggestion for a small change.

I think that `GetAvailableArt` methods should have "arttype" as an optional parameter acting as a filter, and when this filter is not applied (or is blank) all the scraped URLs are returned. This also means returning "arttype" as a results property. This way API consumers can just retrieve all the scraped art URLs for an item without having to make multiple JSON requests.

File: xbmc/interfaces/json-rpc/schema/methods.json [hunk 1]
```diff
@@ -1023,6 +1023,59 @@
...
+      { "name": "arttype", "type": "string", "required": true }
+    ],
+    "returns": {
+      "type": "object",
+      "properties": {
+        "availableart": { "type": "array", "required": true,
+          "items": {
+            "type": "object",
+            "properties": {
+              "url": { "type": "string", "description": "URL to the original image", "required": true },
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/schema/methods.json:1071]: Add field 
"arttype": { "type": "string", "description": "arttype", "required": true },

File: xbmc/interfaces/json-rpc/schema/methods.json [hunk 2]
```diff
@@ -1023,6 +1023,59 @@
...
+    "permission": "ReadData",
+    "params": [
+      {
+        "name": "item", "required": true,
+        "type": [
+          { "type": "object", "properties": { "albumid": { "$ref": "Library.Id", "required": true } }, "additionalProperties": false },
+          { "type": "object", "properties": { "artistid": { "$ref": "Library.Id", "required": true } }, "additionalProperties": false }
+        ]
+      },
+      { "name": "arttype", "type": "string", "required": true }
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/schema/methods.json:1062]: Make "arttype" an optional filter `"required": false`

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 3]
```diff
@@ -663,6 +663,77 @@ JSONRPC_STATUS JSONRPC::CAudioLibrary::GetSources(const std::string& method, ITr
...
+  StringUtils::ToLower(artType);
+
+  CMusicDatabase musicdatabase;
+  if (!musicdatabase.Open())
+    return InternalError;
+
+  CVariant availableart = CVariant(CVariant::VariantTypeArray);
+  for (const auto& artentry : musicdatabase.GetAvailableArtForItem(mediaID, mediaType, artType))
+  {
+    CVariant item = CVariant(CVariant::VariantTypeObject);
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp:725]: `item["arttype"] = artentry.m_aspect;`
May also want to handle empty => "thumb" here, or in `CMusicDatabase::GetAvailableArtForItem`

File: xbmc/music/MusicDatabase.cpp [hunk 4]
```diff
@@ -12228,36 +12228,57 @@ bool CMusicDatabase::GetArtTypes(const MediaType &mediaType, std::vector<std::st
...
+      thumbURL = album.thumbURL;
+  }
+
+  std::vector<CScraperUrl::SUrlEntry> result;
+  for (const auto& urlEntry : thumbURL.GetUrls())
+  {
+    std::string itemArtType = urlEntry.m_aspect;
+    if (itemArtType.empty())
+      itemArtType = "thumb";
+    if (itemArtType == artType)
```
Comment (DaveTBlake) [inline xbmc/music/MusicDatabase.cpp:12280]: if (artType.empty() || itemArtType == artType)

**Review by DaveTBlake — COMMENTED** [review]
Thanks for cleaning up `CMusicDatabase::GetAvailableArtTypesForItem`, as you spotted processing is common now it no longer needs to handle artist fanart separately.

Here is how I see the optional arttype filtering working for music

**Comment by rmrector** [comment]
Yeah! 'arttype' as another list item property is the ticket.

When I thought about returning all of them, I kept going to a map of lists keyed on arttype, which I kept not liking.

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp (hunk 3 above)
Comment (rmrector) [reply xbmc/interfaces/json-rpc/AudioLibrary.cpp:725]: I pushed the empty handling down to the DB function.

**Comment by DaveTBlake** [comment]
Yeap, tested as expected and looks good. Rebase (another version bump sneeked in so v12.2.0 now) and squash and its good to go.

**Comment by rmrector** [comment]
Thanks, I appreciate the quick review and idea.

**Pull request merged**